internal/util/merge/merge3.go

```diff
@@ -0,0 +1,165 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package merge
+
+import (
+	"path/filepath"
+	"strings"
+
+	"github.com/GoogleContainerTools/kpt/pkg/kptfile"
+	"sigs.k8s.io/kustomize/kyaml/kio"
+	"sigs.k8s.io/kustomize/kyaml/kio/filters"
+	"sigs.k8s.io/kustomize/kyaml/kio/kioutil"
+	"sigs.k8s.io/kustomize/kyaml/pathutil"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+const (
+	mergeSourceAnnotation = "config.kubernetes.io/merge-source"
+	mergeSourceOriginal   = "original"
+	mergeSourceUpdated    = "updated"
+	mergeSourceDest       = "dest"
+)
+
+// Merge3 performs a 3-way merge on the original, upstream and
+// destination packages. It provides support for doing this only for
+// the parent package and ignore any subpackages. Whenever the boundaries
+// of a package differs between original, upstream and destination, the
+// boundaries in destination will be used.
+type Merge3 struct {
+	OriginalPath       string
+	UpdatedPath        string
+	DestPath           string
+	MatchFilesGlob     []string
+	MergeOnPath        bool
+	IncludeSubPackages bool
```

Comment:
Are we retaining this flag for update command ?

Reply:
No, we don't expose it to users. But it is useful for this merge implementation.